File: python/example11_4.py
# Example 11.4 Autoregressive Distributed Lag Model
# Almon Lag Model Once More, using data from 1953Q1 - 1961Q4
# Almon Lag (Lags=7, Order=4 End=2): tested and estimated

import pandas as pd
import numpy as np
import statsmodels.api as sm

almon = pd.read_csv("http://web.pdx.edu/~crkl/ceR/data/almon.txt",sep='\s+',nrows=36)
cexp = almon['CEXP']
capp = almon['CAPP']

X = pd.concat([capp,capp.shift(1),capp.shift(2),capp.shift(3),capp.shift(4),
              capp.shift(5),capp.shift(6),capp.shift(7)], axis=1)
X.columns = ['CAPP','CAPP1','CAPP2','CAPP3','CAPP4','CAPP5','CAPP6','CAPP7'] 
Y = pd.concat([cexp.shift(1),cexp.shift(2)],axis=1)
Y.columns = ['CEXP1','CEXP2']

xvar = pd.concat([Y,X],axis=1)
xvar = sm.add_constant(xvar)
xvar = xvar.dropna()
yvar = cexp.iloc[7:]

# unrestricted PDL 7-lags model using OLS
pdl1 = sm.OLS(yvar, xvar).fit()
print(pdl1.summary())

def PDL(q,p):
    Hmat = np.empty((q,p))
    for i in range(q):
        for j in range(p):
            Hmat[i,j]=(i+1)**(j+1)
    Hmat = np.concatenate(([np.zeros(p)],Hmat))
    Hmat = np.concatenate((np.ones(q+1).reshape(-1,1),Hmat),axis=1)
    return(Hmat)

# polynomial lags: restricted, lags=7 order=4
H = PDL(7,4)
Z = X @ H
Z.columns = ['Z0','Z1','Z2','Z3','Z4']
zvar = pd.concat([Y,Z],axis=1)
zvar = sm.add_constant(zvar)
zvar = zvar.dropna()
yvar = cexp.iloc[7:]

pdl2 = sm.OLS(yvar,zvar).fit()
print(pdl2.summary())

# hypothsis testing: check for end-point restrictions
h1 = '(Z0-Z1+Z2-Z3+Z4=0)'                # left-end restriction
h2 = '(Z0+8*Z1+64*Z2+512*Z3+4096*Z4=0)'  # right-end restriction
print(pdl2.wald_test(h1))
print(pdl2.wald_test(h2))
print(pdl2.wald_test([h1,h2]))

# PDL(7,4) with both end-point restrictions
from statsmodels.api import GLM
cons_r = [[0,0,0,1,-1,1,-1,1],
          [0,0,0,1,8,64,512,4096]]
cons_q = [0,0]
pdl3 = sm.GLM(yvar,zvar).fit_constrained((cons_r,cons_q))
print(pdl3.summary())

a_coef = H@pdl3.params[3:]
v = np.array(pdl3.cov_params())
a_se = np.sqrt(np.diag(H @ v[3:,3:] @ H.T))
a_z = a_coef/a_se
res = pd.DataFrame({'coef':a_coef,'std err':a_se,'z':a_z},
                   index=X.columns)
print(res)

File: R/example11_4.R
# Example 11.4 Autoregressive Distributed Lag Model
# Almon Lag Model Once More
# Almon Lag (Lags=7, Order=4)
# Almon Lag (Lags=7, Order=4 End=2): difficult, not estimated, but tested
#
almon<-read.table("http://web.pdx.edu/~crkl/ceR/data/almon.txt",header=T,nrows=60)
cexp<-ts(almon$CEXP,start=c(1953,1),end=c(1961,4),frequency=4)
capp<-ts(almon$CAPP,start=c(1953,1),end=c(1961,4),frequency=4)  

# define lag operator L on one variable
L<-function(x,l) {c(rep(NA,l),x[1:(length(x)-l)])}

H<-function(q,p) {
  Hmat<-matrix(nrow=q,ncol=p)
  for(i in 1:q) {
    for(j in 1:p) {Hmat[i,j]<-i^j}
  }
  Hmat<-rbind(rep(0,p),Hmat)
  Hmat<-cbind(rep(1,q+1),Hmat)
  Hmat
}

Y<-cbind(L(cexp,1),L(cexp,2))
X<-cbind(capp,L(capp,1),L(capp,2),L(capp,3),L(capp,4),L(capp,5),L(capp,6),L(capp,7))
H<-H(7,4)
Z<-X%*%H

# unrestricted 7-lags model
pdl1<-lm(cexp~Y+X)
summary(pdl1)

# polynomial lags: restricted, lags=7 order=4
pdl2<-lm(cexp~Y+Z)
summary(pdl2)

H%*%pdl2$coefficients[4:8]  # using coefficients of Zs
v<-vcov(pdl2)
diag(H%*%v[4:8,4:8]%*%t(H))

# check for end-point restrictions
library(car)
lht(pdl2,"Z1-Z2+Z3-Z4+Z5=0")               # left-end restriction
lht(pdl2,"Z1+8*Z2+64*Z3+512*Z4+4096*Z5=0") # right-end restriction

# check for autocorrelation
dwt(pdl2)
rm(list = ls())
